### PR TITLE
Disallow unknown fields from resources struct

### LINKED ISSUE #297: Do not allow extra fields when validating CRDs
## Expected Behavior

If I typo a field in an optional field in a CRD (i.e. provide an extra field), the validation webhook (or worst case, runtime check) should tell me the CRD is invalid.

## Actual Behavior

"Extra" fields are ignored, for example in the demo pipeline, there was a bug where one of the instances of `inputSourceBindings` was not renamed to `resources`:

https://github.com/knative/build-pipeline/blob/c9341e7f0e5dae53a4c5b1638401b4d65401f1b3/examples/pipeline.yaml#L24-L44

This was ignored (and everything still worked b/c at that time the Pipeline executed in the order it was declared in).

## Steps to Reproduce the Problem

1. Add an extra invalid field to a CRD, for example you can modify [the build-push.yaml example spec](https://github.com/knative/build-pipeline/blob/master/examples/build-task.yaml) to have the extra field `helloIDoNotBelong: foo`:

```yaml
apiVersion: pipeline.knative.dev/v1alpha1
kind: Task
metadata:
  name: build-push
  namespace: default
spec:
  helloIDoNotBelong: foo
  inputs:
...
```

2. Apply the CRD:

```bash
kubectl apply -f examples/build-push.yaml
```

3. Notice that it applies just fine and the extra field is ignored!

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

Before this change, it is possible to passe any additionnal field (in
the yaml/json form) with Tekton resources. This is now disallowed.

Closes #297

cc @bobcatfish 

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) and [TaskRun](../tekton/publish-run.yaml) to build and release this image

# Release Notes

```
Unknown fields in resource YAML/JSON will now be rejected when running "kubectl apply"
```

### CODE REVIEW

File: examples/pipelineruns/pipelinerun.yaml [hunk 1]
```diff
@@ -123,9 +123,9 @@ spec:
       description: Okay this is a hack, but I didn't feel right hard-codeing `-d1` down below
     - name: yamlPathToImage
       description: The path to the image to replace in the yaml manifest (arg to yq)
-    clusters:
-    - name: targetCluster
-      description: Not yet used, kubectl command below should use this cluster
+    #clusters:
```
Comment (vdemeester) [inline examples/pipelineruns/pipelinerun.yaml]: Wasn't sure where this comes from. I am guessing we should remove this altogether most likely :angel:

**Comment by sbwsg** [comment]
This is great. I have definitely been confused by this behaviour before.

/lgtm

File: cmd/webhook/main.go [hunk 2]
```diff
@@ -97,8 +97,8 @@ func main() {
 			v1alpha1.SchemeGroupVersion.WithKind("TaskRun"):          &v1alpha1.TaskRun{},
 			v1alpha1.SchemeGroupVersion.WithKind("PipelineRun"):      &v1alpha1.PipelineRun{},
 		},
-		Logger: logger,
-
+		Logger:                logger,
+		DisallowUnknownFields: true,
```
Comment (bobcatfish) [inline cmd/webhook/main.go]: For my own curiosity: is this a feature that got added somewhere recently, or something we could have been using all along?

File: examples/pipelineruns/pipelinerun.yaml (hunk 1 above)
Comment (bobcatfish) [reply examples/pipelineruns/pipelinerun.yaml]: oh wow! yeah feel free to remove it, this is leftover from a really old version of the spec XD

File: cmd/webhook/main.go (hunk 2 above)
Comment (bobcatfish) [reply cmd/webhook/main.go]: and p.s. if it's not new, do you happen to know why ppl looking at this issue before thought we couldn't easily fix it? (https://github.com/tektoncd/pipeline/issues/297#issuecomment-444568074)

**Comment by sbwsg** [comment]
Looking at this once more, minor suggestion for the release note: At the moment it reads `Unknown field in resources are now disabled`.  I suggest updating this to something a bit more specific like `Unknown fields in resource YAML/JSON will now be rejected when running "kubectl apply"`.  This feels *almost* like a breaking change to me since yamls that used to work now may not in the next version.

File: cmd/webhook/main.go (hunk 2 above)
Comment (vdemeester) [reply cmd/webhook/main.go]: It was added with https://github.com/knative/pkg/pull/338, so ~4month ago :wink:

**Comment by vdemeester** [comment]
Interesting failure :sweat: 

```
I0716 11:40:40.841] >>>> taskrun list-file
I0716 11:40:45.162] Tailing logs for taskrun: list-file...
I0716 12:18:23.709] panic: interface conversion: runtime.Object is nil, not *v1.Pod
I0716 12:18:23.710] 
I0716 12:18:23.710] goroutine 39 [running]:
I0716 12:18:23.710] github.com/tektoncd/pipeline/test/logs/pod.(*Watcher).Start.func1(0x1208620, 0xc000048060, 0x11f89c0, 0xc000416ba0, 0xc0004167b0)
I0716 12:18:23.710] 	/go/src/github.com/tektoncd/pipeline/test/logs/pod/watcher.go:59 +0x19f
I0716 12:18:23.711] created by github.com/tektoncd/pipeline/test/logs/pod.(*Watcher).Start
I0716 12:18:23.711] 	/go/src/github.com/tektoncd/pipeline/test/logs/pod/watcher.go:51 +0x285
I0716 12:18:23.715] exit status 2
I0716 12:18:23.742] >>>> taskrun multiple-build-push-kaniko-run
```

@sbwsg thx :heart_eyes_cat: I'll update :wink:

**Comment by vdemeester** [comment]
```
I0716 12:20:51.287] {"level":"error","logger":"controller.taskrun-controller","caller":"taskrun/taskrun.go:238","msg":"Failed to validate taskrun \"list-file\": invalid input resources: TaskRun's declared resources didn't match usage in Task: Didn't provide required values: [rules]","knative.dev/controller":"taskrun-controller","stacktrace":"github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/taskrun.(*Reconciler).reconcile\n\t/go/src/github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/taskrun/taskrun.go:238\ngithub.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/taskrun.(*Reconciler).Reconcile\n\t/go/src/github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/taskrun/taskrun.go:125\ngithub.com/tektoncd/pipeline/vendor/github.com/knative/pkg/controller.(*Impl).processNextWorkItem\n\t/go/src/github.com/tektoncd/pipeline/vendor/github.com/knative/pkg/controller/controller.go:330\ngithub.com/tektoncd/pipeline/vendor/github.com/knative/pkg/controller.(*Impl).Run.func1\n\t/go/src/github.com/tektoncd/pipeline/vendor/github.com/knative/pkg/controller/controller.go:282"}
```
Interesting that it failed there but not when applying :thinking:  :man_facepalming: 
Preparing a follow-up, as we don't validate *embedded* task specs.

File: examples/pipelineruns/pipelinerun.yaml (hunk 1 above)
Comment (vdemeester) [reply examples/pipelineruns/pipelinerun.yaml]: forgot to update before removing the hold :sweat_smile:

Comment (abergmeier) [reply examples/pipelineruns/pipelinerun.yaml]: Could you perhaps have a test, which actually validates this behavior. Would feel better when not only the happy-paths are tested.

Comment (vdemeester) [reply examples/pipelineruns/pipelinerun.yaml]: @abergmeier yes, working on it (once I land somewhere 🤭)

**Pull request merged**